File: SubsetsII.py
"""
90. Subsets II
Given a collection of integers that might contain duplicates, nums, return all possible subsets.

Note:
Elements in a subset must be in non-descending order.
The solution set must not contain duplicate subsets.
For example,
If nums = [1,2,2], a solution is:

[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]
"""
# O(n * 2 ^ n)
# O(n)
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        res = []
        self.recur(res, nums, [])
        return res
        
    def recur(self, res, nums, cur):
        res.append(cur[:])
        i = 0
        while i < len(nums):
            cur.append(nums[i])
            self.recur(res, nums[i + 1:], cur)
            cur.pop()
            i += 1
            while i < len(nums) and nums[i] == nums[i - 1]:
                i += 1


if __name__ == "__main__":
    print(Solution().subsetsWithDup([1,2,2])

File: Java/SubsetsII.java
import java.util.*;
public class SubsetsII {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> cur = new ArrayList<Integer>();
        recur(cur, res, nums, 0);
        return res;
    }
    
    private void recur(List<Integer> cur, List<List<Integer>> res, int[] nums, int start) {
        res.add(new ArrayList<Integer>(cur));
        int i = start;
        while(i < nums.length) {
            cur.add(nums[i]);
            recur(cur, res, nums, i + 1);
            cur.remove(cur.size() - 1);
            i++;
            while(i < nums.length && nums[i] == nums[i - 1]) {
                i++;
            }
        }
    }

}
